### ISSUE TITLE
Is there any HDL code for this SPI controller?

### ISSUE THREAD

**Issue by Paebbels** [drom/spi#2]
Hello,

is there any HDL code for this SPI controller?

Regards
     Patrick

**Comment by drom** [comment]
Hi Patrick,
No, there is no HDL to commit yet. I am at the specification phase now.
Do you have specific ideas, requirements to the IP? or, plans to contribute?
--Alexey

**Comment by Paebbels** [comment]
Hello,

I was just scanning GitHub for new hardware/HDL repos. I'm one of the maintainers of the [PoC-Library](https://github.com/VLSI-EDA/PoC) and I'm looking if we could cooperate with other repos :).

We have plans for an SPI master, too, but there is no design yet. I would suggest to support all 4 SPI modes and to decouple the SPI master from the specific memory controller(s). Maybe you can divide it into 3-4 modules:
- BusController -> bit level operations like clock generation, data line sampling, input synchronization
- SPIController -> byte level transfers, so it could be used for SPI memories or ADCs, or whatever
- optional GenericSPIMemoryController -> a layer on-top of the byte level controller
- SPIMemoryController -> implementing the vendor specific transaction protocols.

We try to decouple such modules in our PoC Library, so modules and layers can be reused.
Have you already chosen an HDL?

The newest PoC IP cores can be found on the release branch. Maybe, our test infrastructure could be of interest, too. The up coming documentation is hosted at [ReadTheDocs](http://poc-library.readthedocs.io/en/latest/).

Regards
    Patrick

**Comment by drom** [comment]
@Paebbels I like your idea on dividing SPI controller into several blocks / layers. I will start with defining Layer-0 #3, and see what would be the right portion of functionality and module interface.

**Comment by drom** [comment]
@Paebbels I have started PHY design here: #7 I guess it is consistent with "BusController" that you have mentioned above.

**Comment by Paebbels** [comment]
Yes your approach looks good. I added an inline comment at your commit.
